llama-index-core/llama_index/core/memory/simple_composable_memory.py

```diff
@@ -0,0 +1,150 @@
+            formatted_history += (
+                f"\n=====Relevant messages from memory source {ix + 1}=====\n\n"
+            )
+            for m in chat_history:
+                formatted_history += f"\t{m.role.upper()}: {m.content}\n"
+            formatted_history += (
+                f"\n=====End of relevant messages from memory source {ix + 1}======\n\n"
+            )
+
+        formatted_history += DEFAULT_OUTRO_HISTORY_MESSAGE
+        return formatted_history
+
+    def get(self, input: Optional[str] = None, **kwargs: Any) -> List[ChatMessage]:
+        """Get chat history."""
+        return self._compose_message_histories(input, **kwargs)
+
+    def _compose_message_histories(
+        self, input: Optional[str] = None, **kwargs: Any
+    ) -> List[ChatMessage]:
+        """Get chat history."""
+        # get from primary
+        messages = self._primary_memory.get(input=input, **kwargs)
+
+        # get from secondary
+        secondary_histories = []
+        for mem in self._secondary_memory_sources:
+            secondary_history = mem.get(input, **kwargs)
+            if len(secondary_history) > 0:
+                secondary_histories.append(secondary_history)
+
+        # format secondary memory
+        if len(secondary_histories) > 0:
+            single_secondary_memory_str = self._format_secondary_messages(
+                secondary_histories
+            )
+
+            # add single_secondary_memory_str to chat_history
+            if len(messages) > 0 and messages[0].role == MessageRole.SYSTEM:
+                system_message = messages[0].content.split(
+                    DEFAULT_INTRO_HISTORY_MESSAGE
+                )[0]
+                messages[0] = ChatMessage(
+                    content=system_message.strip() + single_secondary_memory_str,
+                    role=MessageRole.SYSTEM,
+                )
+            else:
+                messages.insert(
+                    0,
+                    ChatMessage(
+                        content="You are a helpful assistant."
+                        + single_secondary_memory_str,
+                        role=MessageRole.SYSTEM,
+                    ),
+                )
+        return messages
+
+    def get_all(self) -> List[ChatMessage]:
+        """Get all chat history.
+
+        Uses primary memory get_all only.
+        """
+        return self._primary_memory.get_all()
+
+    def put(self, message: ChatMessage) -> None:
+        """Put chat history."""
+        self._primary_memory.put(message)
+        for mem in self._secondary_memory_sources:
+            mem.put(message)
+
+    def set(self, messages: List[ChatMessage]) -> None:
+        """Set chat history."""
+        self._primary_memory.set(messages)
+        for mem in self._secondary_memory_sources:
+            mem.set(messages)
+
+    def reset(self) -> None:
+        """Reset chat history."""
+        self._primary_memory.reset()
+
+    def reset_all(self) -> None:
```

Comment:
i think that would break the expectation of the interface, since reset on a single memory module would delete everything, but they would need to call reset_all if using a composable memory module

Reply:
That's a fair point. I'm good with foregoing `reset_all` then 👍🏼